code/firmware/rosco_m68k_firmware/blockdev/ata.c

```diff
@@ -316,16 +316,16 @@ static void ata_probe() {
 // This means ata_init **must** be called before any video devices are
 // initialized!
 
-extern void INSTALL_TEMP_BERR_HANDLER(void);
-extern void RESTORE_BERR_HANDLER(void);
 static uint8_t *berr_flag = (uint8_t*)BERR_FLAG;
 
 void ata_init() {
-    INSTALL_TEMP_BERR_HANDLER();
+    INSTALL_TEMP_BERR_HANDLER_C(&&post_write);      // Yes, this is a GCC extension.
+                                                    // No, I do not care. :D
 
     // Disable interrupt
     idereg[ATA_REG_WR_DEVICE_CONTROL] = 0x0002;
 
+post_write:
```

Comment:
Particularly, if the compiler decides to change the stack pointer here or somewhere else that `INSTALL_TEMP_BERR_HANDLER_C` would be called in the future, it wouldn't get restored. Registers could also have unexpected values (especially depending on optimization level).

Reply:
Although I don't love it as a solution, one alternative would be to create a C macro like `PROBE_MEMORY_WRITE_W(addr, value)`, which uses inline asm to probe the assembly way.

This would also allow for the 68000 continuation address to be stored in a register instead of in `SDB_STATUS`. This would require a bit of extra case in assembly code that probes (so I'm not pushing strongly for this, it's just something that came to mind), but it will require fewer memory accesses while probing.

Reply:
Fair point, this _was_ a hack. Although I believe safe _in this specific case_, definitely had potential to create hard to debug problems down the line (the kicker for me was optimisation levels as you mentioned - jumping back to the middle of a function would likely cause all sorts of weirdness that'd be super confusing 😅 ).

I've just moved this whole bit back to assembly.